File: code/Python/unit_tests/test_auxiliary_functions.py
import unittest
import numpy as np
import sys
sys.path.insert(0, "../spiSeMe")
import spiSeMe_distribution_test
import spiSeMe_event_autocorrelation
import spiSeMe_event_cross_correlation

class TestStringMethods(unittest.TestCase):
	def setUp(self):
		self.data_A = np.loadtxt('test_data_A.dat')
		self.data_B = np.loadtxt('test_data_B.dat')
		self.data_Z = [1, 2, 3, 4, 5, 6]

	def test_aux_distrtest_arg(self):
		print("Distribution Test: Intentionally providing bad arguments...")
		p, d = spiSeMe_distribution_test.spiSeMe_distribution_test(self.data_Z, self.data_A)
		self.assertFalse(p)
		self.assertFalse(d)
		p, d = spiSeMe_distribution_test.spiSeMe_distribution_test(self.data_A, self.data_Z)
		self.assertFalse(p)
		self.assertFalse(d)
	def test_aux_distrdtest_run(self):
		print("Distribution Test: Checking output formats...")
		self.p, self.d = spiSeMe_distribution_test.spiSeMe_distribution_test(self.data_A, self.data_A)
		self.assertEqual(self.p, 1.0)
		self.assertEqual(self.d, 0.0)
		self.p, self.d = spiSeMe_distribution_test.spiSeMe_distribution_test(self.data_A, self.data_B)
		self.assertAlmostEqual(self.p, 0.5, delta=0.1)
		self.assertAlmostEqual(self.d, 0.04, delta=0.01)

	def test_aux_autocorr_arg(self):
		print("Autocorrelation: Intentionally providing bad arguments...")
		A, lags = spiSeMe_event_autocorrelation.spiSeMe_event_autocorrelation(self.data_Z, 0.1, 1.0)
		self.assertFalse(A)
		self.assertFalse(lags)
		A, lags = spiSeMe_event_autocorrelation.spiSeMe_event_autocorrelation(self.data_A, -0.1, 1.0)
		self.assertFalse(A)
		self.assertFalse(lags)
		A, lags = spiSeMe_event_autocorrelation.spiSeMe_event_autocorrelation(self.data_A, 0.1, -1.0)
		self.assertFalse(A)
		self.assertFalse(lags)
		A, lags = spiSeMe_event_autocorrelation.spiSeMe_event_autocorrelation(self.data_A, 1.0, 0.1)
		self.assertFalse(A)
		self.assertFalse(lags)
		A, lags = spiSeMe_event_autocorrelation.spiSeMe_event_autocorrelation(self.data_A, 'x', 1.0)
		self.assertFalse(A)
		self.assertFalse(lags)
		A, lags = spiSeMe_event_autocorrelation.spiSeMe_event_autocorrelation(self.data_A, 0.1, 'x')
		self.assertFalse(A)
		self.assertFalse(lags)
	def test_aux_autocorr_run(self):
		print("Autocorrelation: Checking output formats...")
		A, lags = spiSeMe_event_autocorrelation.spiSeMe_event_autocorrelation(self.data_A, 0.1, 1.0)
		self.assertTrue(isinstance(A,np.ndarray))
		self.assertTrue(isinstance(lags,np.ndarray))
		self.assertEqual(A.ndim, 1)
		self.assertEqual(lags.ndim, 1)
		self.assertEqual(A.size, lags.size)
		self.assertEqual(A.size, 10)

	def test_aux_crosscorr_arg(self):
		print("Cross Correlation: Intentionally providing bad arguments...")
		C, lags = spiSeMe_event_cross_correlation.spiSeMe_event_cross_correlation(self.data_Z, self.data_A, 0.1, 1.0)
		self.assertFalse(C)
		self.assertFalse(lags)
		C, lags = spiSeMe_event_cross_correlation.spiSeMe_event_cross_correlation(self.data_A, self.data_Z, 0.1, 1.0)
		self.assertFalse(C)
		self.assertFalse(lags)
		C, lags = spiSeMe_event_cross_correlation.spiSeMe_event_cross_correlation(self.data_A, self.data_B, 'x', 1.0)
		self.assertFalse(C)
		self.assertFalse(lags)
		C, lags = spiSeMe_event_cross_correlation.spiSeMe_event_cross_correlation(self.data_A, self.data_B, 0.1, 'x')
		self.assertFalse(C)
		self.assertFalse(lags)
		C, lags = spiSeMe_event_cross_correlation.spiSeMe_event_cross_correlation(self.data_A, self.data_B, -0.1, 1.0)
		self.assertFalse(C)
		self.assertFalse(lags)
		C, lags = spiSeMe_event_cross_correlation.spiSeMe_event_cross_correlation(self.data_A, self.data_B, 0.1, -1.0)
		self.assertFalse(C)
		self.assertFalse(lags)
		C, lags = spiSeMe_event_cross_correlation.spiSeMe_event_cross_correlation(self.data_A, self.data_B, 1.0, 0.1)
		self.assertFalse(C)
		self.assertFalse(lags)
	def test_aux_crosscorr_run(self):
		print("Cross Correlation: Checking output formats...")
		C, lags = spiSeMe_event_cross_correlation.spiSeMe_event_cross_correlation(self.data_A, self.data_B, 0.1, 1.0)
		self.assertTrue(isinstance(C,np.ndarray))
		self.assertTrue(isinstance(lags,np.ndarray))
		self.assertEqual(C.ndim, 1)
		self.assertEqual(lags.ndim, 1)
		self.assertEqual(C.size, lags.size)
		self.assertEqual(C.size, 21)


if __name__ == '__main__':
    unittest.main()

File: code/C++/unit_tests/test_auxiliary_functions.cpp
#ifndef SPISEME_HEADER
	#define SPISEME_HEADER
	#include "spiSeMe.h"
#endif
#include <fstream>
#include <sstream>

#define CATCH_CONFIG_MAIN
#include "catch.hpp"

TEST_CASE("Test auxiliary functions"){

	// Load data
	std::ifstream input_file_A("test_data_A.dat", std::ifstream::in);
	std::ifstream input_file_B("test_data_B.dat", std::ifstream::in);
	double			temp_iei;
	std::string 		line;
	std::istringstream	linestream;
	std::vector<double>	data_A;
	std::vector<double>	data_B;
	while (std::getline(input_file_A, line)) {
		linestream.str(line);
		linestream >> temp_iei;
		data_A.push_back(temp_iei);
		linestream.clear();
	}
	input_file_A.close();
	while (std::getline(input_file_B, line)) {
		linestream.str(line);
		linestream >> temp_iei;
		data_B.push_back(temp_iei);
		linestream.clear();
	}
	input_file_B.close();

	std::vector<double>	A;
	std::vector<double>	C;
	std::vector<double>	lags;
	std::vector<double>	data_Z(1, 0.0);
	spiSeMe_return_code	err;

	SECTION("Auxiliary test KS-test (arg)") {
		double p, d;
		err = spiSeMe_distribution_test(p, d, data_A, data_Z);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_distribution_test(p, d, data_Z, data_A);
		REQUIRE(err == SSM_BAD_ARGUMENT);
	}

	SECTION("Auxiliary test KS-test (run)") {
		double p, d;
		err = spiSeMe_distribution_test(p, d, data_A, data_B);
		REQUIRE(err == SSM_SUCCESS);
		REQUIRE(fabs(p-0.5) <= 0.1);
		REQUIRE(fabs(d-0.04) <= 0.01);
	}

	SECTION("Auxiliary test Autocorrelation (arg)") {
		err = spiSeMe_event_autocorrelation(A, lags, data_Z, 0.1, 1.0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_event_autocorrelation(A, lags, data_A, -0.1, 1.0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_event_autocorrelation(A, lags, data_A, 0.1, -1.0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_event_autocorrelation(A, lags, data_A, 1.0, 0.1);
		REQUIRE(err == SSM_BAD_ARGUMENT);
	}

	SECTION("Auxiliary test Autocorrelation (run)") {
		err = spiSeMe_event_autocorrelation(A, lags, data_A, 0.1, 1.0);
		REQUIRE(err == SSM_SUCCESS);
		REQUIRE(A.size() == 10);
		REQUIRE(A.size() == lags.size());
	}

	SECTION("Auxiliary test Cross-Correlation (arg)") {
		err = spiSeMe_event_cross_correlation(C, lags, data_Z, data_A, 0.1, 1.0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_event_cross_correlation(C, lags, data_A, data_Z, 0.1, 1.0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_event_cross_correlation(C, lags, data_A, data_B, -0.1, 1.0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_event_cross_correlation(C, lags, data_A, data_B, 0.1, -1.0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_event_cross_correlation(C, lags, data_A, data_B, 1.0, 0.1);
		REQUIRE(err == SSM_BAD_ARGUMENT);
	}

	SECTION("Auxiliary test Cross-Correlation (run)") {
		err = spiSeMe_event_cross_correlation(C, lags, data_A, data_B, 0.1, 1.0);
		REQUIRE(err == SSM_SUCCESS);
		REQUIRE(C.size() == 21);
		REQUIRE(C.size() == lags.size());
	}

}
